### ISSUE TITLE
[HOLD for payment 2022-01-19] App shows blank screen when user denies camera access when validating identity - Reported by @adeel0202

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#6841]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to app 
2. Send money to someone 
3. Click pay with expensify 
4. Deny access to camera 

## Expected Result:
User should be able to verify identity or clear error should display 

## Actual Result:
In iOS, if you deny the camera permission and try to verify identity again, you see the blank screen with an unclear error message.
In Android, if you deny the camera permission and try to verify identity again, you do see the proper message and settings option to allow the permissions manually, however, when you allow the permission in the settings app and come back to the app, you see the same blank screen with the unclear error message. After that, if you try verify identity again, then the camera opens fine.

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Android
- IOS

**Version Number:** 1.1.22-0
**Reproducible in staging?:** Y
**Reproducible in production?:** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43995119/146826784-7e4ce6e7-97bc-4b3f-b571-0871fb0e56a2.mp4

https://user-images.githubusercontent.com/43995119/146826790-a6e4df24-eb99-416a-a4b8-78483a1bdfdb.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @adeel0202
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1639768520380700

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mountiny (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by mountiny** [comment]
@mvtglobally Could you by any chance reproduce this in production so we can make sure this is a regression or persistent issue. Thank you!

**Comment by mvtglobally** [comment]
@mountiny I am able to reproduce in PROD app as well

**Comment by MelvinBot** [comment]
Triggered auto assignment to @bfitzexpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by mountiny** [comment]
@mvtglobally Thank you for double checking.

I will try to look into this, but it in case someone is faster to solve this, I am also making it `External`. Feel free to submit proposals!

**Comment by mountiny** [comment]
@bfitzexpensify 🎅 -bump

**Comment by bfitzexpensify** [comment]
Done! 

Upwork job to resolve issue is [here](https://www.upwork.com/jobs/~01e5328d9e73539fe9) 

Upwork job for reporting is [here](https://www.upwork.com/jobs/~01fae76f5114e0c58a) - @adeel0202

**Comment by cbhushan1995** [comment]
@bfitzexpensify Have you guys tried to observe camera permission whenever the user's application state changes ( background ->  Foreground vice versa ). I think it could be the issue. I can look into this.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Current assignee @mountiny is eligible for the Exported assigner, not assigning anyone new.

**Comment by mountiny** [comment]
@cbhushan1995 I believe the issue might lay in the way how we re-check the camera permissions. Feel free to write a proposal here if you manage to discover what the problem is.

**Comment by cbhushan1995** [comment]
@mountiny is there any test credentials by which i could check existing functionality of app ?

**Comment by mountiny** [comment]
@cbhushan1995 Feel free to refer [the readme files](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#test-accounts) in the repository or ask in the Slack channel as someone will be able to help you faster usually. Essentially, just create your personal account.

Are you in the Slack channel? If not, follow [these instructions](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#asking-questions) and you will get invited 😊

**Comment by phivh** [comment]
Hi, @mountiny
I've just reproduced, there should be a missing implementation from 3rd, seems for iOS they still didn't know how to catch the permission denied and show the page to navigate the user to the setting screen as you saw on Android.
Ref: https://github.com/onfido/react-native-sdk/issues/17#issuecomment-767488275
This is a proposal for now:
1. Add function to check camera permission
2. Observing app state to restart Fido step

**Comment by parasharrajat** [comment]
Thanks for the proposal @phivh .
I have couple of questions and requests.
1. What is the reason behind the blank screen?

>    Add function to check camera permission
 >   Observing app state to restart Fido step
 
 How would you do that? Where will you do that?

**Comment by phivh** [comment]
> Thanks for the proposal @phivh . I have couple of questions and requests.
> 
> 1. What is the reason behind the blank screen?
> 
> > Add function to check camera permission
> > Observing app state to restart Fido step
> 
> How would you do that? Where will you do that?
Hi @parasharrajat 
1. What is the reason behind the blank screen?
There is no cache error message "Onfido.OnfidoFlowError" which is returned from Fido, then it renders a null value, that is why you saw a blank screen.

How would you do that? Where will you do that?
There are 2 ways:
1. We can add one more library camera, it would support us to check the permission camera before starting Fido
2. It should be taking more time - As we can cache error from line 38 at this file: src/components/Onfido/index.native.js to detach it is camera denied then show an alert for user enable the camera permission on iOS, on Android we can create a function to observe app state when the user comes back after enabling the camera.

**Comment by hafizsameed** [comment]
Solution:
If user denies access to the camera then, we can do following steps:
1)we will show a screen with Error message so, that user should what went wrong.
2) On the same screen, there will be a button to give access to the camera (user might have accidentally denied the access so, he/she should not face any problem. they can always provide access to the camera by pressing the button)
3) We will also provide button to go back so, user can go back instead of stuck on that screen if he/she is not willing to provide access.

**Comment by mountiny** [comment]
@phivh and @hafizsameed Both of your proposals are reasonable. 

In general, a good UX practice is to make sure user always knows what state the app is at. If there is some problem, user should know what is wrong. Thus if they deny access to camera and they need access to camera to complete this essential task, we should show them screen which says something like: "Uh-oh, you have not granted us camera access. We need that to complete this verification" and then "Click this button to grant the access" (exact copy would be different).

Then coming back from the settings if required, the camera should work as it will have access, ideally without any other action required by the user.

That is in summary what you have both described. However, if you look over other issues you will notice we would like to get moreless exact code changes required for this to be fixed. Where in code will change which line to achieve this behaviour. It makes the entire process more transparent and we can spot earlier if some proposal is not viable because if something works in high-level does not mean it will actually work implemented.

I realize this requires more time invested into the proposal form your side, but if you are sure your proposal will work and it will be in line with the coding patterns used throughout the codebase, we will most likely pick that proposal! 😊 

Would you be able to describe what exactly need to be changed, and ideally attach video proving these changes work? 

Me and/or Rajat will help you later to polish the code style and iron out any potential problems!

**Comment by phivh** [comment]
Hi @mountiny, I've proposed 2 options above and I want to demonstrate a simple one that catches by the error message, and this is related to existing ways that are currently on the app.

After debugging, there are still missing 2 error messages on iOS and 1 crashing message on android.

So, I can add them into `src/CONST.js`
```
ERROR: {
    ...
    USER_CAMERA_DENINED: 'Onfido.OnfidoFlowError',
    USER_CAMERA_PERMISSION: 'Encountered an error: cameraPermission',
    USER_CAMERA_CONSENT_DENIED: 'Unexpected result Intent. It might be a result of incorrect integration, make sure you only pass Onfido intent to handleActivityResult. It might be due to unpredictable crash or error. Please report the problem to android-sdk@onfido.com. Intent: null \n resultCode: 0',
},
```

Then handle in `src/components/Onfido/index.native.js` at catching error step:
```
.catch((error) => {
    const errorMessage = lodashGet(error, 'message', CONST.ERROR.UNKNOWN_ERROR);
    console.log('errorMessage:', error.message)
    // If the user cancels the Onfido flow we won't log this error as it's normal. In the React Native SDK the user exiting the flow will trigger this error which we can use as
    // our "user exited the flow" callback. On web, this event has it's own callback passed as a config so we don't need to bother with this there.
    if (_.contains([CONST.ONFIDO.ERROR.USER_CANCELLED, CONST.ONFIDO.ERROR.USER_TAPPED_BACK], errorMessage)) {
        this.props.onUserExit();
        return;
    }

    // Added 2 mising errors message with alert to ask user give permission on ios (Onfido related issue: https://github.com/onfido/react-native-sdk/issues/17#issuecomment-767488275)
    // Add 1 missing error message that related to camera user consent on android handle by browser it self (Onfido related issue: https://github.com/onfido/react-native-sdk/issues/11)
    if (_.contains([CONST.ONFIDO.ERROR.USER_CAMERA_PERMISSION, CONST.ONFIDO.ERROR.USER_CAMERA_DENINED, CONST.ONFIDO.ERROR.USER_CAMERA_CONSENT_DENIED], errorMessage)) {
        Alert.alert('"New Expensify" Would Like to Access the Camera ', 'Uh-oh, you have not granted us camera access. We need that to complete this verification', [ 
            {
                text: 'Cancel',
                onPress: () => {
                    this.props.onUserExit();
                }
            },
            {
                text: 'Open Settings',
                onPress: () => {
                    this.props.onUserExit();
                    Linking.openSettings(); 
                }
            },
        ]); 
        return;
    }

    this.props.onError(errorMessage);
});
```

### Demo
#### iOS

https://user-images.githubusercontent.com/5449119/147688615-2d0c3a79-7dd6-48e9-a09a-3c345e7fdcb3.mp4

#### Android

https://user-images.githubusercontent.com/5449119/147688633-5c7801f0-2071-4040-9e98-acdc15c4087b.mp4

Please let me know if there is any concerns from you guys.

**Comment by parasharrajat** [comment]
Thanks for @phivh providing more details. There are a few concerns. 
- Why are we comparing the whole error message? 
- Can we rely on this message for a long time? 

Don't we have an error code or something to identify the error???

You can help me understand this better via linking me to the documentation, if possible.

**Comment by mountiny** [comment]
I agree with @parasharrajat that it might be more future-proof to use some code rather than the exact error message text (in case they would change it). 

However, in general, this is a very good proposal. Great job @phivh! Can you first help us understand the Onfido reports better (as Rajat mentioned, link the documentation) before we move on? Thank you very much!

**Comment by phivh** [comment]
Hi @mountiny @parasharrajat 
> Why are we comparing the whole error message?
Don't we have an error code or something to identify the error???

This was also my concern when I looked into the error handling step (the code was implemented by Marc Glasser from the team), then I thought it could be a reason somewhere. So after looking deep into Onfido SDK I found there is quite constraint for error handling.

- There is an issue related to still opening at least 1 year ago.[`https://github.com/onfido/react-native-sdk/issues/16`](https://github.com/onfido/react-native-sdk/issues/16) with no response from Onfido dev team.
- This is code implemented from Onfido SDK - ReactNative, it seems simple and only one `error_code = config_error` is configured https://github.com/onfido/react-native-sdk/blob/3af6460de542a730ccedc5a6ecf52d3639269aa4/js/Onfido.js#L67  . Otherwise, it will return `error_code = error` and seems not to specify to us to handle a specific case. 
- This is on the Readme [https://github.com/onfido/react-native-sdk#4-failure-response](https://github.com/onfido/react-native-sdk#4-failure-response) only one `error_code = config_error` mentioned.

To summarize, the only way that we can handle error response is only based on error message.

> Can we rely on this message for a long time?

- I would say no because it could be changed by Onfido SDK, if it happens we also need to upgrade accordingly.
- If this is for the bug fix I would use it. If we want to use the long term it could be an enhancement / improvement by creating functionality to request camera permission by Expensify app at TOS screen when clicking on the button `Continue` before Onfido start.

**Comment by parasharrajat** [comment]
Ok. Looking at this, I have second thoughts. 

I don't think we can rely on error messages especially when they can break anytime. We will require to create an integration test to test that error messages are still the same which is overhead.

**But I don't see any other straight way of doing it. I think we can adopt this solution until we have better options available.** From this perspective, @phivh proposal looks good to me. 

cc: @mountiny What do you think?

**Comment by mountiny** [comment]
@parasharrajat @phivh Great write up Brian, thank you for this. I agree this is by far not the ideal solution, but if it is the best we have right now and it solves this problem, let's do it. 

I believe if we cover this scenario in our QA steps this should never get to production before we notice something has changed. Additionally, we add the integration test, that would be even safer.

@phivh Welcome on board! I will assign you to this issue and feel free to make the PR! Thank you very much. I could create a follow-up issue to create the integration tests so we catch this (not sure how easy that will be). You might be interested in working on that one too.

@phivh Can you please apply for this job on Upwork? And then let @bfitzexpensify know by commenting here you have done so and Ben will hire you! Thank you Rajat as well.

**Comment by MelvinBot** [comment]
📣 @phivh You have been assigned to this job by @mountiny!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by phivh** [comment]
@mountiny
Thanks! I'm preparing PR.
@parasharrajat 
Who should I request to code review?

Hi @bfitzexpensify, I've just applied for the job on Upwork.

**Comment by parasharrajat** [comment]
@phivh it would be automatic and you don't have permission to request review.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @davidcardoza (`Waiting for copy`), see https://stackoverflow.com/c/expensify/questions/7025/ for more details.

**Comment by mountiny** [comment]
@davidcardoza Hi! 👋 We have two new messages being shown through this PR. For Onfido KYC flow, in case user denies access to camera, we need to re-prompt the user since the camera access is essential. So when they try to verify, this prompt will show:

![image](https://user-images.githubusercontent.com/36083550/147837522-24589237-90a8-4944-be72-7645af919d23.png)

Now the copy is:

cameraRequestTitle: `'"New Expensify" Would Like to Access the Camera'`
cameraRequestMessage: `'Uh-oh, you have not granted us camera access. We need that to complete this verification'`

which was just a quick proposal, so we can rewrite it completely as you wish. Thank you very much for your help

**Comment by mountiny** [comment]
Waiting for a copy.

**Comment by mountiny** [comment]
Friendly bump @davidcardoza 👋

**Comment by davidcardoza** [comment]
Can you change "We need that to complete this verification" to "We need access to complete verification." All looks good otherwise.

**Comment by mountiny** [comment]
Of course! Thank you David! 🙌

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.27-1 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/6969


If no regressions arise, payment will be issued on 2022-01-19. :confetti_ball:

**Comment by bfitzexpensify** [comment]
All payments issued, we're all done here
